### PR TITLE
use undici.request for httpClient

### PR DESCRIPTION
**Pull request by pushrax**
according to the undici docs this method is much faster than fetch https://github.com/nodejs/undici/tree/main#benchmarks

my preliminary performance testing agrees with this, notably better combined body_read_time+deserialize_time when the process is under load

this is passing tests and should work, but could use:

- [x] test that gunzip actually works (the test suite responses are too small to trigger it)
- [ ] stream gunzip? edit: decided not to
    - advantages of streamed gunzip:
        - in theory it's better to pipeline the cpu work with the io
    - advantages of oneshot gunzip:
        - can time the body read and gunzip separately
        - maybe drive i/o faster to buffer the whole response sooner
        - fewer context switches to C
- [x] separate metric for gunzip / json?
- [x] reduce lint errors

### CODE REVIEW

File: src/httpClient.ts [hunk 1]
```diff
@@ -270,3 +276,35 @@ function make_request_timing(
...
+async function consumeResponseText(response: Dispatcher.ResponseData): Promise<{
+  body_text: string;
+  body_read_end: number;
+}> {
+  if (response.headers["content-encoding"] == "gzip") {
+    let body_buffer = await response.body.arrayBuffer();
+    let body_read_end = performance.now();
+
+    let gunzip_buffer = await gunzip_async(body_buffer);
+    let body_text = gunzip_buffer.toString(); // is there a better way?
```
Comment (pushrax) [inline src/httpClient.ts:303]: undici-fetch uses [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) but idk if it's any better

File: src/httpClient.ts [hunk 2]
```diff
@@ -1,7 +1,11 @@
-import type { Response } from "undici";
+import type { Dispatcher } from "undici";
 import { fetch, Agent } from "undici";
 import pako from "pako";
 import { version } from "../package.json";
+import { gunzip } from "node:zlib";
+import { promisify } from "node:util";
+
+const gunzip_async = promisify(gunzip);
```
Comment (sirupsen) [inline src/httpClient.ts:8]: Why is this required? Presumably it's still blocking underneath, so what's the reason to not just call `gunzip`?

File: src/httpClient.ts [hunk 3]
```diff
@@ -125,7 +136,7 @@ class DefaultHTTPClient implements HTTPClient {
     }
```
Comment (sirupsen) [inline src/httpClient.ts:136]: Should we use the `node:zlib` bindings instead of pako to encode the request? ^

File: src/httpClient.ts [hunk 4]
```diff
@@ -134,12 +145,13 @@ class DefaultHTTPClient implements HTTPClient {
       error = null;
       request_start = performance.now();
       try {
-        response = await fetch(url.toString(), {
-          method,
+        response = await this.agent.request({
```
Comment (sirupsen) [inline src/httpClient.ts:148]: Did you verify that this still pools properly? 👍🏻

File: src/httpClient.ts [hunk 5]
```diff
@@ -160,32 +172,29 @@ class DefaultHTTPClient implements HTTPClient {
...
-        });
+        error = new TurbopufferError(
+          message ?? `http error ${response.statusCode}`,
+          {
+            status: response.statusCode,
+          },
+        );
       }
       if (
         error &&
```
Comment (sirupsen) [inline src/httpClient.ts:200]: Might be worth adding the retries to the requestTiming, to ensure those are extremely infrequent (as they'll add a lot of latency). A `retryTime` timing, for example

File: src/httpClient.ts (hunk 1 above)
Comment (sirupsen) [reply src/httpClient.ts:303]: Doubt it... 

As per your comment, you can `JSON.parse` buffer directly, but doesn't seem like that makes any difference, [see code](https://gist.github.com/sirupsen/603855eef0fbe9c576beddabc4266a60)

File: src/httpClient.ts [hunk 6]
```diff
@@ -270,3 +276,35 @@ function make_request_timing(
...
+
+async function consumeResponseText(response: Dispatcher.ResponseData): Promise<{
+  body_text: string;
+  body_read_end: number;
+}> {
+  if (response.headers["content-encoding"] == "gzip") {
+    let body_buffer = await response.body.arrayBuffer();
+    let body_read_end = performance.now();
+
+    let gunzip_buffer = await gunzip_async(body_buffer);
```
Comment (sirupsen) [inline src/httpClient.ts:302]: > stream gunzip?

I feel like this is not worth it over the benefits of one-shot, especially the timing we can add here

>  separate metric for gunzip / json?

yes; you can add a combined if you prefer, but I feel like that's just body_read

File: src/httpClient.ts (hunk 2 above)
Comment (pushrax) [reply src/httpClient.ts:8]: it's callback based, not blocking, promisify makes it possible to await it

gunzipSync is blocking and would be really bad to use on a production server

File: src/httpClient.ts (hunk 3 above)
Comment (pushrax) [reply src/httpClient.ts:136]: sure we may as well

File: src/httpClient.ts (hunk 4 above)
Comment (pushrax) [reply src/httpClient.ts:148]: yes it definitely does

File: src/httpClient.ts (hunk 6 above)
Comment (pushrax) [reply src/httpClient.ts:302]: we already have deserialize_time which is the combined time, question is whether to make multiple timings for the two steps

File: src/httpClient.ts [hunk 7]
```diff
@@ -1,7 +1,11 @@
-import type { Response } from "undici";
+import type { Dispatcher } from "undici";
 import { fetch, Agent } from "undici";
 import pako from "pako";
 import { version } from "../package.json";
+import { gunzip } from "node:zlib";
```
[the remark this answers is not in the document]
Comment (pushrax) [reply src/httpClient.ts:5]: this file is becoming node-centric so we could get rid of pako, sure, there was no logic

File: src/httpClient.ts (hunk 6 above)
Comment (sirupsen) [reply src/httpClient.ts:302]: Yes I definitely think we should

File: src/httpClient.ts (hunk 5 above)
Comment (pushrax) [reply src/httpClient.ts:201]: for a followup, I think we're adding enough metrics rn, and we know the issue isn't retries because we see all the time in the old body_read_time metric

**Review by sirupsen — APPROVED** [review]
👍🏻 to non-streamed gzip

Only thing I'll leave to your discretion is whether you want to add `request_serialization_time` and `request_compression_time`

And based on our conversation, whether you want to stick with zlib or use pako

**Comment by sirupsen** [comment]
I guess for non-streamed GZIP, it's intuition, so a benchmark will always be better if it's not a huge PITA.

**Comment by pushrax** [comment]
- sticking with zlib, should be more efficient and worst case with pako is worse
- sticking with oneshot gunzip for now since measurement is clear, can optimize it later
- keeping the additional metric load to the most important ones for now
